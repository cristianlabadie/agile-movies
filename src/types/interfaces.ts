export interface ResponseError {
  message: string
  statusCode: number
}
export interface ResponseLogin {
  data: Data
}

export interface Data {
  payload: Payload
  user: User
}

export interface Payload {
  refresh_token: string
  token: string
  type: string
}

export interface User {
  email: string
  firstName: string
  lastName: string
}

export interface ResponseJWTError {
  error: string
  message: string
  statusCode: number
}

export interface ResponseUserMe {
  data: Data
}

export interface UseInfo {
  createdAt: Date
  email: string
  firstName: string
  id: number
  lastName: string
  password: null
  username: string
}

export interface ResponseMovies {
  imageBaseUrl: string
  data: Movie[]
}

export interface Movie {
  adult: boolean
  backdrop_path: string
  genre_ids: number[]
  id: number
  original_language: OriginalLanguage
  original_title: string
  overview: string
  popularity: number
  poster_path: string
  release_date: Date
  title: string
  video: boolean
  vote_average: number
  vote_count: number
}

export enum OriginalLanguage {
  En = 'en',
  Ja = 'ja',
  No = 'no',
  Tl = 'tl'
}

export interface ResponseActors {
  imageBaseUrl: string
  data: Actor[]
}

export interface Actor {
  adult: boolean
  gender: number
  id: number
  known_for_department: KnownForDepartment
  name: string
  original_name: string
  popularity: number
  profile_path: null | string
  cast_id: number
  character: string
  credit_id: string
  order: number
}

export enum KnownForDepartment {
  Acting = 'Acting',
  Crew = 'Crew'
}
