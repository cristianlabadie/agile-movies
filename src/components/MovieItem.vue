<template>
  <div
    class="absolute z-10 bottom-0 top-0 w-full h-full bg-gradient-to-b from-transparent from-40% via-[#151836]/50 to-[#151836]/80"
  ></div>

  <div
    class="absolute left-0 top-0 bottom-0 w-full h-full group-hover:scale-110 transition-scale duration-1000 ease-in-out opacity-90 bg-blend-luminosity bg-center bg-cover bg-no-repeat -z-10"
    :class="{
      'animate-pulse bg-slate-500 transition-all': loadingImg
    }"
    :style="loadingImg ? '' : getBgUrl"
  ></div>

  <div
    class="relative group-hover:cursor-pointer select-none flex flex-col gap-1 p-6 text-lg z-20 justify-end h-full"
    :class="{ 'min-h-96': extend }"
  >
    <h2
      class="text-sm font-semibold line-clamp-2 text-balance text-white mb-4 uppercase"
      :class="{ 'md:text-3xl': extend }"
      id="title"
    >
      {{ movie.title }}
    </h2>

    <h3
      class="text-xs line-clamp-2 font-semibold text-sky-200/80 -mt-4 mb-4 sm:mb-0"
      v-if="movie.overview"
      id="overview"
    >
      {{ movie.overview }}
    </h3>
  </div>
</template>
<script setup lang="ts">
import { ref, type PropType, onMounted, computed } from 'vue'
import type { Movie } from '@/types/interfaces'
import { useMoviesStore } from '@/stores/movies'

const props = defineProps({
  movie: {
    type: Object as PropType<Movie>,
    required: true
  },
  extend: {
    type: Boolean,
    default: false
  }
})

const movieStore = useMoviesStore()

const imgMovie = new Image()
const loadingImg = ref(true)

imgMovie.onload = () => {
  loadingImg.value = false
}

const imageToShow = computed(() => {
  return props.extend
    ? movieStore.getBasePath + props.movie.backdrop_path
    : movieStore.getBasePath + props.movie.poster_path
})

onMounted(() => {
  imgMovie.src = imageToShow.value
})

const getBgUrl = computed(() => {
  return {
    backgroundImage: `url(${imgMovie.src})`
  }
})
</script>
