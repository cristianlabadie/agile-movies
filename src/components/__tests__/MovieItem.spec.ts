import { describe, it, expect } from 'vitest'
import { mount } from '@vue/test-utils'
import { createTestingPinia } from '@pinia/testing'
import MovieItem from '../MovieItem.vue'
import { OriginalLanguage } from '@/types/interfaces'
import { vi } from 'vitest'

describe('Pruebas en MovieItem component', () => {
  const movie = {
    id: 1,
    title: 'Spiderman',
    overview: 'Spiderman overview',
    backdrop_path: '/spiderman.jpg',
    poster_path: '/spiderman.jpg',
    vote_average: 5.5,
    release_date: new Date('2021-09-01'),
    adult: false,
    genre_ids: [],
    original_language: OriginalLanguage.En,
    original_title: '',
    popularity: 1,
    video: false,
    vote_count: 1
  }

  const wrapper = mount(MovieItem, {
    global: {
      plugins: [
        createTestingPinia({
          createSpy: vi.fn,
          initialState: {
            movies: {
              movies: [movie]
            }
          }
        })
      ]
    },
    props: {
      movie: movie
    }
  })

  it('Debe de renderizar el componente correctamente', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('Debe de mostrar el título de la película', () => {
    const title = wrapper.find('#title')
    expect(title.text()).toBe(movie.title)
  })

  it('Debe de mostrar el overview de la película', () => {
    const overview = wrapper.find('#overview')
    expect(overview.text()).toBe(movie.overview)
  })
})
