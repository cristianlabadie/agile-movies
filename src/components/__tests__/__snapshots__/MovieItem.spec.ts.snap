// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Pruebas en MovieItem component > Debe de renderizar el componente correctamente 1`] = `
"<div class="absolute z-10 bottom-0 top-0 w-full h-full bg-gradient-to-b from-transparent from-40% via-[#151836]/50 to-[#151836]/80"></div>
<div class="absolute left-0 top-0 bottom-0 w-full h-full group-hover:scale-110 transition-scale duration-1000 ease-in-out opacity-90 bg-blend-luminosity bg-center bg-cover bg-no-repeat -z-10 animate-pulse bg-slate-500 transition-all" style=""></div>
<div class="relative group-hover:cursor-pointer select-none flex flex-col gap-1 p-6 text-lg z-20 justify-end h-full">
  <h2 class="text-sm font-semibold line-clamp-2 text-balance text-white mb-4 uppercase" id="title">Spiderman</h2>
  <h3 class="text-xs line-clamp-2 font-semibold text-sky-200/80 -mt-4 mb-4 sm:mb-0" id="overview">Spiderman overview</h3>
</div>"
`;
